### PR TITLE
3112392040 create csv upload page

### PR DESCRIPTION
**Pull request by Nuuggs**

### CODE REVIEW

File: staff/urls.py [hunk 1]
```diff
@@ -17,6 +17,7 @@
     path('basics/batches/<int:batch_id>/sections/', section.ListView.as_view(), name='section_list'),
     path('basics/batches/<int:batch_id>/sections/<int:section_id>/', section.DetailView.as_view(), name='section_detail'),
     path('coupons/', coupon.ListView.as_view(), name='coupon_list'),
+    path('coupons/csv-upload/', coupon.CsvUploadView.as_view(), name='coupon_csv_upload'),
```
Comment (samklq) [inline staff/urls.py:20]: did some reading online - wonder if it makes sense to just POST to `coupons/`. but then we'll still need another route for rendering the form. alternatively we rename this to be `coupons/new/batch/` or `coupons/new_many/` for consistency with the other routes?

https://www.codementor.io/blog/batch-endpoints-6olbjay1hd
https://developer.zendesk.com/api-reference/ticketing/users/users/#create-or-update-many-users
https://stackoverflow.com/questions/28596688/rest-api-bulk-create-or-update-in-single-request
https://stackoverflow.com/questions/28638071/update-an-entire-resource-collection-in-a-rest-way

File: staff/forms/csv_upload.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        for index, row in enumerate(csv_file):
+            row_data = str(row.decode('utf-8')).split(",")
+            first_name = row_data[0]
+            # replace to remove csv formatting for new row
+            email = row_data[1].replace('\\r\\n', '')
+            if index == 0:
+                if "first_name" not in first_name and "email" not in email:
+                    self.add_error(
+                        'csv_file',
+                        'The file you uploaded requires the headers "first_name" and "email" !'
```
Comment (samklq) [inline staff/forms/csv_upload.py:33]: pulled code and uploaded a file that didn't have these headers, but didn't see this error rendered on screen. can take a look? prob wanna add test coverage for this.

File: staff/templates/coupon/csv_upload_success.html [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+                <tr>
+                    <td>{{ row.first_name }}</td>
+                    <td>{{ row.email }}</td>
+                </tr>
+                {% endfor %}
+            </tbody>
+        </table>
+        <div class="row">
+            <div class="d-flex flex-row py-5">
+            <a type="link" href="/staff/coupons/" role="button" class="btn btn-outline-secondary px-4">Back to Coupons List</a>
```
Comment (samklq) [inline staff/templates/coupon/csv_upload_success.html:25]: for consistency with rest of app, don't need to capitalise everything - "Back to coupons" will do

File: staff/tests/views/coupon/test_coupon_csv_upload.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+    assert response.status_code == HttpResponse.status_code
+    assert 'coupon/csv_upload.html' in (template.name for template in response.templates)
+
+def test_csv_upload_success_renders_page_with_list_of(logged_in_existing_user):
+    test_file_path = "./staff/tests/forms/csv_files/correct_test_file.csv"
+    csv_file = open(test_file_path, 'r')
+    content = csv_file.read()
+    uploaded_file = SimpleUploadedFile(name=csv_file.name, content=bytes(content, 'utf-8'), content_type="multipart/form-data")
+    assert_result = [{'first_name': 'tester', 'email': 'test1@test.com'}, {'first_name': 'tester2', 'email': 'test2@gmail.com'}]
```
Comment (samklq) [inline staff/tests/views/coupon/test_coupon_csv_upload.py:49]: this variable is not used anywhere else, think you can just leave this array in the assert

File: staff/templates/coupon/coupon_generation.html [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+            <label>Upload a .csv file</label>
+            {{ coupon_generation_form.csv_file }}
+            <small class="text-danger">
+                {{ coupon_generation_form.csv_file.errors|striptags }}
+            </small>
+        </div>
+        <div class="row">
+            <div class="d-flex flex-row py-5">
+            <a type="link" href="/staff/coupons/" role="button" class="btn btn-outline-secondary px-4">Cancel</a>
+            <button class="btn btn-primary mx-2 px-4" type="submit">Upload file</button>
```
Comment (samklq) [inline staff/templates/coupon/coupon_generation.html:18]: this should be "Generate coupons"?

File: staff/tests/forms/test_coupon_generation_form.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+class TestCouponGenerationForm:
+    def test_empty_form_is_invalid(self):
+        coupon_generation_form = CouponGenerationForm(data={})
+
+        outcome = coupon_generation_form.is_valid()
+
+        assert outcome is False
+
+    def test_file_uploaded_field_accepts_csv(self):
```
Comment (samklq) [inline staff/tests/forms/test_coupon_generation_form.py:16]: maybe the test name should reflect that we're testing for validation on incorrect file name?

File: staff/tests/forms/test_coupon_generation_form.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+                    content=bytes('test content', 'utf-8'),
+                    content_type="multipart/form-data"
+                )
+            }
+        )
+
+        outcome = coupon_generation_form.is_valid()
+
+        assert outcome is False
+        assert not coupon_generation_form.__dict__['files']['csv_file'].name.endswith('.csv')
```
Comment (samklq) [inline staff/tests/forms/test_coupon_generation_form.py:30]: is this necessary? might be better to test for the error message instead.

File: staff/urls.py [hunk 8]
```diff
@@ -17,6 +17,7 @@
     path('basics/batches/<int:batch_id>/sections/', section.ListView.as_view(), name='section_list'),
     path('basics/batches/<int:batch_id>/sections/<int:section_id>/', section.DetailView.as_view(), name='section_detail'),
     path('coupons/', coupon.ListView.as_view(), name='coupon_list'),
+    path('coupons/new/batch/', coupon.CsvUploadView.as_view(), name='coupon_coupon_generation'),
```
Comment (samklq) [inline staff/urls.py:20]: a more consistent naming would be `coupon_new_batch`. if you think it makes sense, can change the form to CouponBatchForm too. sorry my bad - i realise CouponGenerationForm doesn't reflect the 'batch' element well. could be easily misread as being for generating 1 coupon.

File: staff/views/coupon.py [hunk 9]
```diff
@@ -106,3 +109,35 @@ def post(self, request, coupon_id):
                 'coupon_form': coupon_form,
             }
         )
+
+class CsvUploadView(LoginRequiredMixin, View):
```
Comment (samklq) [inline staff/views/coupon.py:113]: update the view class name to reflect NewBatchView or something?

File: staff/views/coupon.py [hunk 10]
```diff
@@ -106,3 +109,35 @@ def post(self, request, coupon_id):
...
+        form = CouponGenerationForm(request.POST, request.FILES)
+
+        if not form.is_valid():
+            return render(request, 'coupon/coupon_generation.html', {'coupon_generation_form': form})
+
+        csv_file = form.cleaned_data.get('csv_file')
+        csvreader = csv.DictReader(codecs.iterdecode(csv_file, 'utf-8'))
+        csv_rows = []
+        for row in csvreader:
+            print('!! row in csv_file !!', row)
```
Comment (samklq) [inline staff/views/coupon.py:134]: remove this print statement pls! tq!

File: staff/views/coupon.py (hunk 9 above)
Comment (samklq) [reply staff/views/coupon.py:113]: need to update the test file name as well

File: staff/templates/coupon/coupon_generation.html (hunk 5 above)
Comment (Nuuggs) [reply staff/templates/coupon/coupon_generation.html:18]: can, but because the label is "upload a .csv file" that's why i didn't change the button

**Pull request merged**